common/rate-limiter/src/rate_limit.rs

```diff
@@ -0,0 +1,446 @@
+/// # Features
+/// ## Keys
+/// The token bucket takes any key as long as it's hashable.  This should allow it to apply to
+/// many applications that need rate limiters.
+///
+/// ## Bucket sizes and Rates
+/// ### Defaults
+/// There are defaults for bucket size and fill rate, which will apply to unknown keys.
+///
+/// ### Refill Interval
+/// Buckets are refilled automatically at an interval.  To do this synchronously, it calculates the
+/// number of intervals that have passed.  This is done synchronously and in the future may be done
+/// asynchronously.
+///
+pub struct TokenBucketRateLimiter<Key: Eq + Hash + Clone + Debug> {
+    buckets: RwLock<HashMap<Key, Arc<Mutex<Bucket>>>>,
+    default_bucket_size: NonZeroUsize,
+    default_fill_rate: NonZeroUsize,
+    open: bool,
+}
+
+impl<Key: Eq + Hash + Clone + Debug> TokenBucketRateLimiter<Key> {
+    pub fn new(default_bucket_size: NonZeroUsize, default_fill_rate: NonZeroUsize) -> Self {
+        assert!(
+            default_bucket_size >= default_fill_rate,
+            "Bucket size must be greater than or equal to fill rate"
+        );
+        Self {
+            buckets: RwLock::new(HashMap::new()),
+            default_bucket_size,
+            default_fill_rate,
+            open: false,
+        }
+    }
+
+    /// Used for testing and to not have a rate limiter
+    pub fn open() -> Self {
+        let max = NonZeroUsize::new(std::usize::MAX).unwrap();
+        Self {
+            buckets: RwLock::new(HashMap::new()),
+            default_bucket_size: max,
+            default_fill_rate: max,
+            open: true,
+        }
+    }
+
+    /// Retrieve bucket, or create a new one
+    pub fn bucket(&self, key: Key) -> Arc<Mutex<Bucket>> {
+        if !self.open {
+            self.bucket_inner(key, |size, rate| {
+                Arc::new(Mutex::new(Bucket::new(size, rate)))
+            })
+        } else {
+            self.bucket_inner(key, |_, _| Arc::new(Mutex::new(Bucket::open())))
+        }
+    }
+
+    fn bucket_inner<F: FnOnce(NonZeroUsize, NonZeroUsize) -> Arc<Mutex<Bucket>>>(
+        &self,
+        key: Key,
+        bucket_create: F,
+    ) -> Arc<Mutex<Bucket>> {
+        // Attempt to do a weaker read lock first, followed by a write lock if it's missing
+        // For the common (read) case, there should be higher throughput
+        let maybe_value = { self.buckets.read().get(&key).cloned() };
+
+        if let Some(value) = maybe_value {
+            value
+        } else {
+            let size = self.default_bucket_size;
+            let rate = self.default_fill_rate;
+            // Write in a bucket, but make sure again that it isn't there first
+            self.buckets
+                .write()
+                .entry(key)
+                .or_insert_with(|| bucket_create(size, rate))
+                .clone()
+        }
+    }
+
+    /// Garbage collects a single key, if we know what it is
+    pub fn garbage_collect_key(&self, key: &Key) -> bool {
```

Comment:
maybe we should call this maybe_garbage_collect_key?

Reply:
Yeah, or `try_garbage_collect_key`, except the try interface usually gives an `Option`.  Thoughts?